src/main/scala/io/opentargets/etl/backend/target/Ensembl.scala

```diff
@@ -1,23 +1,21 @@
 package io.opentargets.etl.backend.target
 
 import com.typesafe.scalalogging.LazyLogging
-import io.opentargets.etl.backend.spark.Helpers.{nest, safeArrayUnion}
-import io.opentargets.etl.backend.target.TargetUtils.transformColumnToIdAndSourceStruct
+import io.opentargets.etl.backend.spark.Helpers._
+import io.opentargets.etl.backend.target.TargetUtils._
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.types.{IntegerType, LongType}
 import org.apache.spark.sql.{Column, DataFrame, Dataset, SparkSession}
 
 case class Ensembl(id: String,
                    biotype: String,
                    approvedName: String,
-                   alternativeGenes: Array[String],
+                   alternativeGenes: Option[Array[String]],
```

Comment:
@JarrodBaker question, what is the behaviour when someone encodes to a case class a data frame and one of the arrays is null instead of empty?

Reply:
If the array is `null` in the Dataframe you get a null field in the Dataset, but all rows are preserved.

```
case class idArray(id: String, maybeArray: Array[String])

val data = Seq(
  Row(1, Array()),
  Row(2, null),
  Row(3, Array("cat"))
)
val schema = List(
  StructField("id", IntegerType, true),
  StructField("maybeArray", ArrayType(StringType), true)
)
val df = ss.createDataFrame(ss.sparkContext.parallelize(data), StructType(schema))

df.show(false)

val asCaseClass = df.as[idArray]

assert(asCaseClass.count() == 3)
```
Output of `asCaseClass.show(false)`:
```
+---+----------+
|id |maybeArray|
+---+----------+
|1  |[]        |
|2  |null      |
|3  |[cat]     |
+---+----------+
```

Reply:
@JarrodBaker great, and when the Option is used then, how does this code behave then?

Reply:
I don't quite follow. Going from `Dataset[Row]` to Dataset[idArray]` as in the example above doesn't change the underlying representation in Spark. Consider:
```
df.printSchema
asCaseClass.printSchema
``` 
yields: 
```
root
 |-- id: integer (nullable = true)
 |-- maybeArray: array (nullable = true)
 |    |-- element: string (containsNull = true)
root
 |-- id: integer (nullable = true)
 |-- maybeArray: array (nullable = true)
 |    |-- element: string (containsNull = true)
```

 If I define the case class where the array is optional and convert the Dataframe to that and examine the schema it is still the same:

```
case class idOptionArray(id: String, maybeArray: Option[Array[String]])
val asOptionCC = df.as[idOptionArray]
asOptionCC.printSchema
```
Shows:
```
root
 |-- id: integer (nullable = true)
 |-- maybeArray: array (nullable = true)
 |    |-- element: string (containsNull = true)
```
If you wanted to use the case class in a UDF or operate on them in native Scala I can see that there would be benefits of having the option type, but we only use the case class conversion as a) a sanity check that all the expected columns are present and of the right type and b) as documentation so that we don't need to dig into the code itself to figure out what is available on the return type.

Reply:
@JarrodBaker Imagine you want to keep processing the array
```
val r = asCaseClass.map(e => e.maybeArray.map(_.toLowerCase))
// it wont fail here until you perform any further action and produce something

@ r.show 
21/09/16 15:04:23 ERROR Executor: Exception in task 2.0 in stage 12.0 (TID 38)
java.lang.NullPointerException
	at scala.collection.mutable.ArrayOps$ofRef$.length$extension(ArrayOps.scala:204)
	at scala.collection.mutable.ArrayOps$ofRef.length(ArrayOps.scala:204)
	at scala.collection.SeqLike.size(SeqLike.scala:108)
	at scala.collection.SeqLike.size$(SeqLike.scala:108)
	at scala.collection.mutable.ArrayOps$ofRef.size(ArrayOps.scala:198)
	at scala.collection.IndexedSeqLike.sizeHintIfCheap(IndexedSeqLike.scala:103)
	at scala.collection.IndexedSeqLike.sizeHintIfCheap$(IndexedSeqLike.scala:103)
	at scala.collection.mutable.ArrayOps$ofRef.sizeHintIfCheap(ArrayOps.scala:198)
	at scala.collection.mutable.Builder.sizeHint(Builder.scala:80)
	at scala.collection.mutable.Builder.sizeHint$(Builder.scala:79)
	at scala.collection.mutable.ArrayBuilder.sizeHint(ArrayBuilder.scala:25)
	at scala.collection.TraversableLike.builder$1(TraversableLike.scala:282)
	at scala.collection.TraversableLike.map(TraversableLike.scala:285)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:198)
	at ammonite.$sess.cmd14$.$anonfun$r$1(cmd14.sc:1)

```
it will result in an exception later when it is performed. In the case of having an Option is easy to infer the lack of future problems
```
@ case class idArray(id: String, maybeArray: Option[Array[String]]) 
defined class idArray

@ val asCaseClass = df.as[idArray] 
asCaseClass: Dataset[idArray] = [id: int, maybeArray: array<string>]

@ val r = asCaseClass.map(e => e.maybeArray.map(_.map(_.toUpperCase))) 
r: Dataset[Option[Array[String]]] = [value: array<string>]

@ r.show(false) 
+-----+
|value|
+-----+
|[]   |
|null |
|[CAT]|
+-----+
```